### ISSUE TITLE
If the DW can be applied to softmax classifier?

### ISSUE THREAD

**Issue by Icecream-blue-sky** [strongwolf/DW#15]
In the code,  DW use sigmoid classifier by default. And I apply it to logo detection(num of categories is 352）. However, I find that there are a lot of FP( same position, different category). I wonder if it is due the use of sigmoid classifier. How could change the DW to use softmax classifier?
![alfaromeo5](https://user-images.githubusercontent.com/26771037/175901711-ecbc8af7-15dd-4220-a07a-7179b03d26b4.jpg)

**Comment by strongwolf** [comment]
The majority of current detectors use sigmoid-based classifiers.  Since the formulations of the two classifiers are different,  it is not direct to apply DW on softmax classifiers. I suggest you can use sigmoid classifiers.

**Comment by Icecream-blue-sky** [comment]
But isn't the principle of the two the same? They both can give different weights to positive and negative samples.

**Comment by strongwolf** [comment]
> But isn't the principle of the two the same? They both can give different weights to positive and negative samples.

The two almost have no difference in the final AP if no weighting scheme is used. Suppose the score and logit of a positive anchor with label j is s_j and z_j,  s_j has an influence on the gradient with respect to z_j in both classifiers and s_j has (no) influence on the gradient with respect to z_k(k!=j) in softmax(sigmoid) classifiers. When we change the weight loss of a positive anchor in softmax, the gradient with respect to all classes will be changed which is not what we want in DW.  Sigmoid-based classifier is more flexible to do sample weighting in a finer level.

**Comment by Icecream-blue-sky** [comment]
> > But isn't the principle of the two the same? They both can give different weights to positive and negative samples.
> 
> The two almost have no difference in the final AP if no weighting scheme is used. Suppose the score and logit of a positive anchor with label j is s_j and z_j, s_j has an influence on the gradient with respect to z_j in both classifiers and s_j has (no) influence on the gradient with respect to z_k(k!=j) in softmax(sigmoid) classifiers. When we change the weight loss of a positive anchor in softmax, the gradient with respect to all classes will be changed which is not what we want in DW. Sigmoid-based classifier is more flexible to do sample weighting in a finer level.

Thanks! So do you have any suggestion to remove these occurrences of FP(same postition, different categories)？

**Comment by Icecream-blue-sky** [comment]
Sorry, I'm little confused about the caculation process of cls Loss of DW, can you give a specific example？For example, if a positive anchor's predicted cls score is [0.3,0.5], the IoU is 0.6, what's the cls loss?
